### ISSUE TITLE
False positive on NO_PROXY parsing can lead to not able to connect

### ISSUE THREAD

**Issue by cthulu** [request/request#1524]
I've hit this issue which is very similar (if not the same) as #1294. It happened to me on 2.54.0 and 2.48.0 but strangely, 2.49.0 was a version which seems to work properly.

As mentioned in #1294 in my case the NO_PROXY contained a host which was exactly 1 character longer than my actual destination host (which should have been accessed through the PROXY). It was however tried to connect directly, which resulted to no connection.

My suspicion is that this is related to some refactoring like https://github.com/request/request/commit/b205b73960b40a1edaa24f4ccede9b676c07cbf2 change but that's only a guess.

**Comment by jbirch-atlassian** [comment]
Hey there,

*tl;dr: probably not a bug*

I'm Cthulu's colleague (There's a sentence I never thought I'd get to say), and while I can't give you access to our environment where we first noticed this, I can distil it down to a small test case of the _opposite_ behaviour (That is, something isn't proxied when it should be)

In the following snippet, we set HTTPS_PROXY to something that isn't a proxy (namely, http://example.com), and try and request something that isn't in the NO_PROXY list. Thus, it should attempt to be proxied, and *fail*.

```
[jbirch@pyke ~/scratch/request-1524]$ cat test_regression.js 
var request = require('request');

var target_url = 'https://support.atlassian.com';
var r = request.get(target_url, function (err, res) {
    console.log(err || res.statusCode);
});

[jbirch@pyke ~/scratch/request-1524]$ npm install request@2.48.0
...
[jbirch@pyke ~/scratch/request-1524]$ npm ls | head -2
/home/jbirch/scratch/request-1524
└─┬ request@2.48.0

[jbirch@pyke ~/scratch/request-1524]$ HTTPS_PROXY='http://example.com' NO_PROXY='.12345678.atlassian.com' node test_regression.js 
200

^^ Ruh Roh.

[jbirch@pyke ~/scratch/request-1524]$ rm -rf node_modules/
[jbirch@pyke ~/scratch/request-1524]$ npm install request@2.49.0
...
[jbirch@pyke ~/scratch/request-1524]$ npm ls | head -2
/home/jbirch/scratch/request-1524
└─┬ request@2.49.0

[jbirch@pyke ~/scratch/request-1524]$ HTTPS_PROXY='http://example.com' NO_PROXY='.12345678.atlassian.com' node test_regression.js 
{ [Error: tunneling socket could not be established, cause=Parse Error] code: 'ECONNRESET' }

^^ Whew
```

However, I can't seem to replicate this with 2.54.0:

```
[jbirch@pyke ~/scratch/request-1524]$ rm -rf node_modules/
[jbirch@pyke ~/scratch/request-1524]$ npm install request@2.54.0
...
[jbirch@pyke ~/scratch/request-1524]$ npm ls | head -2
/home/jbirch/scratch/request-1524
└─┬ request@2.54.0

[jbirch@pyke ~/scratch/request-1524]$ HTTPS_PROXY='http://example.com' NO_PROXY='.12345678.atlassian.com' node test_regression.js 
{ [Error: tunneling socket could not be established, cause=Parse Error] code: 'ECONNRESET' }

^^ As expected.
```

This also works on master (faddddb18d at time of writing), so there's another piece of our puzzle missing here, and it's probable this doesn't affect the request library. Interestingly enough, while '.12345678.atlassian.com' broke the proxying in the old version, '123456789.atlassian.com' did not - I wonder if there might be gremlins in the canonicalisation logic for NO_PROXY hosts that indicate a domain, rather than a specific host. Gremlins may well have been taken care of already :)

**Comment by riemers** [comment]
I am on 2.61.0 and tried 2.60.0, whenever i set no_proxy it will still go over the proxy. I've managed to not have this issue on a mac, but on windows this issue still remains. Removing http_proxy from the environment solves the issue but it should just listen to no_proxy. Working around this now, but thought i should mention it. Sometimes it just doesn't act as its suppose to do with no_proxy.
